doc/manual.rst

```diff
@@ -5453,6 +5468,145 @@ powerful programming construct that still suffices. So the "check list" is:
 (4) Else: Use a macro.
 
 
+Alias Declaration
+=================
+
+An alias declaration creates a symbol that is an alias (alternative name)
+for another symbol and can be used transparently in its place. Any symbol can
+be aliased: templates, modules, types, etc; even overloaded symbols can be
+aliased (see `ast.nkClosedSymChoice`).
+The closest equivalent would be D's [alias](https://dlang.org/spec/declaration.html#alias).
+Examples:
+
+.. code-block:: nim
+  
+  when defined(customEcho): alias myecho = myechoImpl
+  else: alias myecho = echo
+  myecho "foo: ", 12
+  doAssert myecho is proc
+
+  alias
+    system2 = system # module
+    SSInt = system2.SomeSignedInt # fully qualified
+    dollar = `$` # overloaded operator
+
+Note that aliases are resolved upon use, creating a transparent abstraction
+(unlike a typedef declaration for types):
+
+.. code-block:: nim
+
+  type Foo = object
+  type Foo2 = Foo
+  alias Foo3 = Foo
+  doAssert $Foo2 == "Foo2" # `nkTypeDef` is not transparent
+  doAssert $Foo3 == "Foo" # alias is transparent
+
+Aliases can only be used for symbols, not expressions:
+
+.. code-block:: nim
+
+    :test: "nim c $1"
+    :status: 1
+
+  import os
+  # these are all invalid
+  alias foo1 = 1+2
+  alias foo2 = typeof(1+2)
+  alias foo3 = walkDir(".")
```

Comment:
Really? That sucks.

Reply:
this is not a real limitation as explained in the next section which explains how to do all of those.
This is by design, otherwise you'd end up with inherent ambiguities, eg:

```nim
when true:# D20200705T215455
  import std/lambdas
  template bar(): untyped =
    static: echo "in bar" # executed only once
    alias echo
  alias echo2 = bar()
  echo2(1, @[1,2])
  echo2()
```
with current design (and current implementation https://github.com/nim-lang/Nim/pull/11992 albeit with syntax `const echo2 = alias2 bar()` for now), this works and is not ambiguous; `bar()` is executed once and resovles to an alias of echo which can then be used and is indistinguishable from `echo`.

if `alias foo3 = walkDir(".")` were supported, or more generally if `alias foo = expr()` were supported with meaning of calling `expr()` each time `foo` is called, this would become ambiguous (or at very least error prone and with complex edge cases) as shown in above code:
with `alias echo2 = bar()`, it would be ambiguous whether `bar()` is called once and resolved to a symbol immediately (as in current semantics) or whether `foo` would resolve to a calling `expr()` each time it's called. 

So I'm instead taking the conservative route and avoiding all those ambiguities.